### DISCUSSION TITLE
Jewels: Display mod weights

### DISCUSSION THREAD

**Discussion by AnnanFay** [Lailloken/Exile-UI#109] [Ideas & Feedback]
Not sure how useful this is in general, but thought it would be cool to see mod weights on basic jewels.

Maybe it's useful for other items, but jewels stand out the most since ilvl doesn't matter and there are no tiers. They are also something that is easy to reroll at the vendor.

Mod weight is also often slightly indicative of how good a mod is.

![suggestion_L](https://user-images.githubusercontent.com/766048/195992573-756b55f9-7d37-436e-a958-af1ca6bc0372.png)

**Comment by Lailloken** [comment]
Hmm, sounds like a good idea. I'll do some research into finding datamined resources and/or how to datamine them myself. If I figure this out, I can add displaying weights as an optional setting for the item-info feature. Thanks for the suggestion.

**Comment by Lailloken** [comment]
Hey there, finally found some time and data resources to pursue your suggestion. Going through the resources for initial planning, I found some aspects that fundamentally influence how this feature will be implemented, and I need some feedback on what to do:

1. I would replace the tiers altogether and instead use weighting
2. As far as I can tell, there are 9 different weight-values (ranging from 100 to 500, in increments of 50), and their estimated distribution (quick parse through the data, needs to be double-checked) is:
    - 2 x 100, 2 x 150, 6 x 200, 39 x 250, 14 x 300, 23 x 350, 7 x 400, 6 x 450, 72 x 500 (all these include overlapping weights that depend on the jewel type, e.g. max life % on crimson = 500, and 350 on the others)
    - I want to group them together and use the same highlighting as with normal mod-tiers, but I'm unsure how to do it. Example: T1 = 100 - 200, T2 = 250 + 300, etc.
3. Some mods can have vastly different weightings depending on the jewel type, which makes highlighting inconsistent and confusing. Just by glancing over the data, I have already found cases where a mod can have a weighting of 250 or 500.

Essentially, the problem I'm facing is that it seems more difficult to add meaningful visualization or glance-value to the quality of jewels. Some weights don't reflect the mod's usefulness/usability (max life % has 500 on crimson jewels), and others differ vastly (possibly being T2 or T5, depening on the jewel type if tiered as mentioned above).

**Reply by AnnanFay** [reply]
Hi, glad to see progress and that it's feasible.

I wrote this yesterday, but it ended up being a bit of a ramble and unfocused.

Leaving it as was below, but in summary: I don't see an easy way of artificially creating tiers (power) from mod weights (rarity), and any solution which displays weights as if they are tiers will cause issues. I think raw weights with either new colour tiers or linear gradient between two configured colours is a good quick solution which avoids confusion of people conflating weights with tiers. Making the config option very clearly 'display mod crafting weights' will also help.

---------------

A similar but related issue this made me think of applies to other items. STR shields are 25% likely for a mod to roll as +life (2..3% the max tier) while INT shields are only 16% likely to it as life (1.4% the max tier). The mods have the same weight, just different pool sizes. This makes a tier 1 life role on an int shield more rare. More valuable? Maybe. At least it would be interesting if there was a way of seeing this.

Re: Jewels:

### Raw weight values (50, 250, 500, etc.)
Pros:

* Simple to implement
* Doesn't mix up weight (rarity) with tiers (power), they are correlated but displaying the same could confuse people

Cons:

* Will need to add more colours. Though item tiers also go beyond 6 I guess people don't care of difference between lower tiers \*
* Inconsistent across base items - I guess there could be two numbers, weight on this item and normalised across base items

### Normalised to 'Chance of getting a mod on a random jewel' (average across 3 types)

Pros:

* Consistent values across items within a class

Cons:

* More colouring issues because of fractional values - I assume we end up with more than 9 values if average weights
* How to deal with prismatic jewels, though maybe you can use these weights actually instead of average
* Maybe I want to know that a mod is rare on a specific base item (the INT shield case above)

### Binning

Pros:

* Easier at a glance to decide if something is trash
* Easier to configure distinct colouring

Cons:

* With fractional tiers it can make differences look larger than they are - With integer bin widths 2.999 and 3 look different while 2.01 and 2.99 look the same. Mainly an issue if your data is not uniformly spread as a cluster around a bin-edge causes issues.

### Gradients

Pros:

* Faster to configure colours if you only allow top-bottom range colours
* No loss in information when displaying fractional percent change of getting mod
* Easy to adapt to new situations (3.20 may completely changes mod weights so they go from 100 to 3000 - since the weights are arbitrary)

Cons:

* Gradients between some colours are difficult / look bad
* Green in general - need to use something like lab, but not sure if autohotkey has that built in

[\*side note: might be interesting to display mod level instead of tier. It sometimes seems as though 90% of the tier 1 mods I see are physical attack leech because it only has 1 tier on rings at level 50, or 'of Rejuvenation' the level-8 tier-1 life gain on kill. Also even in mods with many tiers, a tier 5 added physical is a level 5 mod while tier 5 max mana is a level 60 mod, quite different power levels across same tier. Though with levels you have even more issues with colouring and would need to implement a gradient of some type]

**Reply by Lailloken** [reply]
Thanks for your very detailed feedback. I'll think about these considerations when I have more time, but here are some things I can comment on right now:

> I think raw weights with either new colour tiers or linear gradient between two configured colours is a good quick solution which avoids confusion of people conflating weights with tiers

Yeah, I worded that poorly. I was trying to find a way to fit the weights into the existing 'slots' -- not equate them to tiers -- in order to avoid having to create separate ones with new colors and such. It's mostly laziness but there are also UX considerations (I don't want to overwhelm users with a bloaty/meaty section in the settings menu).
<br>

> The mods have the same weight, just different pool sizes. This makes a tier 1 life role on an int shield more rare. More valuable? Maybe. At least it would be interesting if there was a way of seeing this.

I could consider tackling that once I'm more comfortable working with the databases. From an outsider's perspective, they are quite messy and unintuitive. It's a single 22MB JSON-file with all the item-mods in the game, even retired ones (which are merely tagged with things such as "old_do_not_use_"). I'll see how things go with jewels, but I'll have to build custom databases derived from the official ones either way. If I find efficient ways to reliably automate that, I might expand weights to other item classes.
<br>

> might be interesting to display mod level instead of tier

I can take this into consideration as an optional toggle, but this also depends on how things go with the databases.

**Comment by Lailloken** [comment]
Small update on my progress (things are mainly just tacked on for testing, so the UI is incoherent):

![image](https://user-images.githubusercontent.com/61888437/202755345-2b6b22a5-4153-44a2-aa13-affa80862fa8.png)

The databases are quite a nightmare in terms of repurposing/reverse-engineering, so I decided to push your suggestion with ilvl-requirements up in priority. It's because creating a database just for jewels is not 100% reliable since the data related to jewels is not explicitly tagged as such, i.e. I'll first create a database for all item-mods instead in order to have a reliable database that potential features can access in the future.

**Comment by Lailloken** [comment]
Alright, things are slowly coming together (system and databases seem to be fully functional for weights and level-requirements):

![image](https://user-images.githubusercontent.com/61888437/202910526-5ebe93f7-6920-488f-b061-f463832d1292.png)
<br>

I still have to decide which approach to take for visual representation and highlighting, but right now I'm leaning towards using color gradients/intensity. I think that's a solution that doesn't invoke associations of good or bad when it's merely based on weights (max life % being 500 springs to mind). I'll also have to weigh up raw values and percentages (or other forms of abstractions) and see what feels the most intuitive or pleasant to use.
***

Some things came up while working on ilvl-requirements, and I need your thoughts:
1. I don't think it's feasible to show the levels **in addition to** the tiers -- optional or not

   - a third column (as in the screenshot in my earlier comment) would essentially make the tooltip a 3x6 matrix of information that would not be easily digestible at a glance: individual red/cyan highlighting, tiering, lvl-requirements; all across 6 or more lines

2. a toggle to replace tiers with lvl-requirements and to highlight mods based on that could potentially lead to extreme cases where ilvl and tier/strength are heavily mismatched, e.g. attack speed on gloves:

   - there are only four tiers, and tier 1 and 2 are fairly close in terms of effect/strength (11-13% vs 14-16%)
   - but their lvl-requirements are 22 vs 76, which doesn't appropriately represent the power-gap and would also be very far apart in any kind of ordering/visualization system
   - such discrepancies would manifest themselves in misleading visual representation and may lead to good mod-rolls being overlooked

3. as a result, I was thinking of implementing ilvl-requirements this way instead:

   - the toggle would say something along the lines of 'also display ilvl-requirements'
   - the colored highlighting would remain unchanged, i.e. based on tiers, but the individual cells would display the ilvl instead
   - an item's potential could still be easily determined at a glance by checking the overall color distribution
   - a closer look at the column would then confirm or disprove the initial assessment because outliers/discrepancies within numbers arranged in a column stand out quite a bit

4. (I just thought of this while typing:) the extreme case would be implementing some sort of score- or grade-system that would factor in all the parameters and spit out a summary or single value with which to decide if an item is worth keeping.

**Comment by AnnanFay** [comment]
Looks good

One question I just thought of looking at the screenshot. You can remove the `H` by combining rows? So bleed chance + bleed duration are in the same table row, but each on their own line. It would also make it more clear when you have items with multiple hybrid mods which ones are linked together.

----

1. I'm not sure. If optional then the user is opting in to less digestible data. Ideally you might have >5 different bits of information and the user can toggle any / all of them. Though I always lean towards showing too much over too little.

    [A method of template processing would be interesting, `{tier}/{max_tier} | {ilvl}`, where `|` would be parsed into table columns to allow full configuration. But that is going *way* out of scope.]

3. I don't see a way of avoiding issues other than showing both at the same time here. Seeing it as `attack speed 12% | 2 | 22` you know it's high tier within the stat but likely going to be very common on items because of the required ilvl. If you don't show two colours (tier-colour with ilvl) then it forces you to read every ilvl - because there is no way of knowing if you need to read the ilvl you must always read it.

4. Difficult to say. Showing both as colours you would see a row with a bad colour and good colour - then know to look closer. In most cases you would see a row with 2 bad, or 2 good colours and would just scan over.

    I guess the key issue here is that some mods you care more about tier while others you care more about ilvl. One other option is to normalise tiers based on available tiers. So tier 3 resistance becomes 3/8 = 37 while tier 3 attack speed becomes 3/4 = 75. With scale being arbitrary 100. For cases where mod doesn't scale to high levels you would likely just hide it (mark as ignore colour). But breaks if there are mods where value are not linear.

5. Funnily enough I do something similar to this for searching my stash. But in a very clunky way through exporting data to google sheets using the looty extension. Works okay for finding some things but quite a few edge cases, and export method means no access to mod-count. One thing I regret not doing is clean split between 'max attack score', 'max spell score', 'total defence score'. Currently it's a per item-class score, global score and mostly broken damage scores. Ping me on Discord/Reddit and I'll share access if interested (AnnanFay#0634)

![image](https://user-images.githubusercontent.com/766048/203254410-fc0e351b-9ae1-46f2-ac34-0edf22f2e849.png)

**Reply by Lailloken** [reply]
> You can remove the `H` by combining rows?

Yes, I'll do that. It was a temporary solution when I started working on the item-info tooltip, and I wanted to overhaul it when finalizing the UI but then somehow forgot about it.
<br>

> One other option is to normalise tiers based on available tiers.

That's an interesting idea, but the databases don't include the available number of tiers of a mod. I just skimmed through them in order to see what would be required to determine that number. I think I see a way to do this, but maybe that's something for a future release. This is only the second time I'm working with JSON data -- on top of that, these databases are huge -- and I think it's better to take iterative steps. Fortunately, it looks like the required data could neatly fit into the structure I chose for my custom databases, so I wouldn't have to restructure that again if I were to implement some sort of normalization.
<br>

> If optional then the user is opting in to less digestible data

> I don't see a way of avoiding issues other than showing both at the same time here.

> Showing both as colours you would see a row with a bad colour and good colour - then know to look closer.

Yeah, I was playing around with different layouts/configurations yesterday to see how things look in practice. My initial impressions are in line with the points you raise, and I agree with the opting-in aspect. I guess I was too afraid that people would blindly tick every checkbox in the settings and then be confused by the amount of information presented to them.

Do you think the six colors implemented right now would be sufficient to visualize ilvl-requirements? Off the top of my head, I would group them together like this (using the default colors):
- lime: 83 to 86 (T16+)
- green: 78 to 82 (red maps)
- yellow: 73 to 77 (yellow maps)
- orange: 67 to 72 (white maps)
- red: 61 to 66 (acts 9 and 10)
- dark red: 60 and below

I also just had the idea to implement a 'league-start mode' or multiple profiles that would shift the top-tier color up or down the stack, something like 'acts', 'early atlas', 'end-game'. This would better visualize how good an item is at certain points of progression. Or maybe an adaptive mode that uses the client.txt log to set the T1-color to, for example, `{current area-level} +- 2`. EDIT: This mode would hide mod-tiers and only show ilvl-requirements.
<br>

> I do something similar to this for searching my stash.

Thanks for sharing and the offer. Once I have finished the current tasks, I'll see if this kind of idea is feasible for my tooltip and how far I get.

**Reply by AnnanFay** [reply]
Sounds good to me. There is no good answer to a colour palette limit. You will always find someone who wants more, though 6 sounds fine to me. I would say that whatever is chosen it should be possible to choose different colours for tiers vs ilvls.

**Comment by Lailloken** [comment]
I just released v1.28.9 which includes jewel weights and ilvl-requirements. The latter is customizable in color, but I could theoretically also make the lvl-thresholds customizable if necessary. Thanks again for your suggestion, and feel free to share more of your ideas (they're really good) and feedback if you think something should be changed/improved. In the near future, I'll also look into the other things that came up here.

**Reply by AnnanFay** [reply]
Looks awesome! Thanks for adding :)
